src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+//SPDX-License-Identifier: None
+pragma solidity ^0.8.0;
+
+import "src/interfaces/IMarket.sol";
+import "src/interfaces/IERC20.sol";
+import "src/interfaces/IDola.sol";
+import "src/interfaces/ITransformHelper.sol";
+import {Ownable} from "lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+import {ReentrancyGuard} from "lib/openzeppelin-contracts/contracts/security/ReentrancyGuard.sol";
+
+// Accelerated leverage engine
+contract ALE is Ownable, ReentrancyGuard {
+    error CollateralNotSet();
+    error MarketNotSetForCollateral(address collateral);
+    error SwapFailed();
+    error DOLAInvalidBorrow();
+    error DOLAInvalidRepay();
+    error InvalidProxyAddress();
+
+    // 0x ExchangeProxy address.
+    // See https://docs.0x.org/developer-resources/contract-addresses
+    address payable public exchangeProxy;
+
+    IDola public immutable dola;
+
+    struct Market {
+        address market;
+        address helper;
+        address collateral;
```

Comment:
Is there a reason we can't just call the market to get the collateral?

Reply:
I thought it was more efficient gas wise to store it when setting a market so we don't have to make an external call to get it